File: npbloom-web/src/__tests__/content/positioned/positioning.test.ts
import { describe, expect, it } from 'vitest';
import {
  applyNodePositionsToTree, CoordsInTree, idMap, PlotCoordsOffset, PositionedTerminalNode, set, sortNodesByXCoord,
  StringSlice, TreeCoordsOffset, TreeXRange, UnpositionedBranchingNode, UnpositionedFormerlyBranchingNode,
  UnpositionedTerminalNode, UnpositionedTree
} from 'npbloom-core';
import mockStrWidth from '../../__mocks__/mockStrWidth';

describe('Node positioning', () => {
  const treeWithTerminalNodes = new UnpositionedTree(
    'Noun verbs.',
    idMap({
      'a': new UnpositionedTerminalNode('N', new TreeCoordsOffset(0, 0), new StringSlice(0, 4)),
      'b': new UnpositionedTerminalNode('V', new TreeCoordsOffset(5, 0), new StringSlice(5, 10)),
    }),
    new PlotCoordsOffset(0, 0),
  );

  const treeWithTriangleNodes = new UnpositionedTree(
    'Noun verb phrases.',
    idMap({
      'a': new UnpositionedTerminalNode('N', new TreeCoordsOffset(0, -10), new StringSlice(0, 4)),
      'b': new UnpositionedTerminalNode('VP', new TreeCoordsOffset(0, 0), new StringSlice(5, 17), true),
    }),
    new PlotCoordsOffset(0, 0),
  );

  const treeWithStrandedNodes = new UnpositionedTree(
    'Noun verbs.',
    idMap({
      'c': new UnpositionedFormerlyBranchingNode('S', new TreeCoordsOffset(0, 0), treeWithTerminalNodes.nodes),
    }),
    new PlotCoordsOffset(0, 0),
  );

  const treeWithBranchingNodes = new UnpositionedTree(
    'Noun verbs.',
    idMap({
      'a': new UnpositionedTerminalNode('N', new TreeCoordsOffset(0, 0), new StringSlice(0, 4)),
      'b': new UnpositionedTerminalNode('V', new TreeCoordsOffset(5, 0), new StringSlice(5, 10)),
      'c': new UnpositionedBranchingNode('S', new TreeCoordsOffset(0, 0), set(['a', 'b'])),
    }),
    new PlotCoordsOffset(0, 0),
  );

  const treeWithBranchingAndTriangleNodes = new UnpositionedTree(
    'Noun verb phrases.',
    idMap({
      'a': new UnpositionedTerminalNode('N', new TreeCoordsOffset(0, -10), new StringSlice(0, 4)),
      'b': new UnpositionedTerminalNode('VP', new TreeCoordsOffset(0, 0), new StringSlice(5, 17), true),
      'c': new UnpositionedBranchingNode('S', new TreeCoordsOffset(0, 0), set(['a', 'b'])),
    }),
    new PlotCoordsOffset(0, 0),
  );

  it('positions terminal nodes', () => {
    const result = applyNodePositionsToTree(mockStrWidth, treeWithTerminalNodes);
    expect(result.node('a').position).toStrictEqual(new CoordsInTree(18, -2));
    expect(result.node('b').position).toStrictEqual(new CoordsInTree(62, -2));
  });

  it('positions terminal nodes with triangles', () => {
    const result = applyNodePositionsToTree(mockStrWidth, treeWithTriangleNodes);
    expect(result.node('a').position).toStrictEqual(new CoordsInTree(18, -12));
    expect(result.node('b').position).toStrictEqual(new CoordsInTree(79.5, -20));
  });

  it('positions stranded nodes', () => {
    const result = applyNodePositionsToTree(mockStrWidth, treeWithStrandedNodes);
    expect(result.node('c').position).toStrictEqual(new CoordsInTree(40, -42));
  });

  it('positions branching nodes', () => {
    const result = applyNodePositionsToTree(mockStrWidth, treeWithBranchingNodes);
    expect(result.node('c').position).toStrictEqual(new CoordsInTree(40, -42));
  });

  it('positions branching and triangle nodes', () => {
    const result = applyNodePositionsToTree(mockStrWidth, treeWithBranchingAndTriangleNodes);
    expect(result.node('c').position).toStrictEqual(new CoordsInTree(48.75, -60));
  });

  it('positions triangle vertices', () => {
    const result = applyNodePositionsToTree(mockStrWidth, treeWithTriangleNodes);
    const triangleNode = result.node('b') as PositionedTerminalNode;
    expect(triangleNode.triangle).toStrictEqual(new TreeXRange(40, 119));
  });

  it('sorts node IDs by X coordinate, given a tree', () => {
    expect(sortNodesByXCoord(mockStrWidth, treeWithBranchingNodes, set(['b', 'a']))).toStrictEqual(['a', 'b']);
    expect(sortNodesByXCoord(mockStrWidth, treeWithBranchingNodes, set(['a', 'c']))).toStrictEqual(['a', 'c']);
  });

  it('measures tree width', () => {
    const result = applyNodePositionsToTree(mockStrWidth, treeWithBranchingAndTriangleNodes);
    expect(result.width).toBe(123);
  });
});

File: npbloom-core/src/commonTest/kotlin/content/positioned/PositioningTest.kt
package content.positioned

import content.EntitySet
import content.StringSlice
import content.YAlignMode
import content.unpositioned.*
import mockStrWidth
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals

class PositioningTest {
    private val treeWithTerminalNodes = UnpositionedTree(
        id = "O29",
        sentence = "Noun verbs.",
        nodes = EntitySet(
            UnpositionedTerminalNode("a", "N", TreeCoordsOffset.ZERO, StringSlice(0, 4)),
            UnpositionedTerminalNode("b", "V", TreeCoordsOffset(5.0, 0.0), StringSlice(5, 10)),
        ),
        offset = PlotCoordsOffset.ZERO,
    )

    private val treeWithTriangleNodes = UnpositionedTree(
        id = "oqYi7",
        sentence = "Noun verb phrases.",
        nodes = EntitySet(
            UnpositionedTerminalNode("a", "N", TreeCoordsOffset(0.0, -10.0), StringSlice(0, 4)),
            UnpositionedTerminalNode("b", "VP", TreeCoordsOffset.ZERO, StringSlice(5, 17), true),
        ),
        offset = PlotCoordsOffset.ZERO,
    )

    private val treeWithStrandedNodes = UnpositionedTree(
        id = "7NI69oA",
        sentence = "Noun verbs.",
        nodes = EntitySet(
            UnpositionedFormerlyBranchingNode("c", "S", TreeCoordsOffset.ZERO, treeWithTerminalNodes.nodes),
        ),
        offset = PlotCoordsOffset.ZERO,
    )

    private val treeWithBranchingNodes = UnpositionedTree(
        id = "BJcX",
        sentence = "Noun verbs.",
        nodes = EntitySet(
            UnpositionedTerminalNode("a", "N", TreeCoordsOffset.ZERO, StringSlice(0, 4)),
            UnpositionedTerminalNode("b", "V", TreeCoordsOffset(5.0, 0.0), StringSlice(5, 10)),
            UnpositionedBranchingNode("c", "S", TreeCoordsOffset.ZERO, setOf("a", "b")),
        ),
        offset = PlotCoordsOffset.ZERO,
    )

    private val treeWithBranchingAndTriangleNodes = UnpositionedTree(
        id = "926mr",
        sentence = "Noun verb phrases.",
        nodes = EntitySet(
            UnpositionedTerminalNode("a", "N", TreeCoordsOffset(0.0, -10.0), StringSlice(0, 4)),
            UnpositionedTerminalNode("b", "VP", TreeCoordsOffset.ZERO, StringSlice(5, 17), true),
            UnpositionedBranchingNode("c", "S", TreeCoordsOffset.ZERO, setOf("a", "b")),
        ),
        offset = PlotCoordsOffset.ZERO,
    )

    private val treeWithVerticallyAlignedNode = UnpositionedTree(
        id = "926mr",
        sentence = "Noun verbs.",
        nodes = EntitySet(
            UnpositionedTerminalNode("a", "N", TreeCoordsOffset.ZERO, StringSlice(0, 4), yAlignMode = YAlignMode.Top),
            UnpositionedTerminalNode("d", "V", TreeCoordsOffset.ZERO, StringSlice(5, 10)),
            UnpositionedBranchingNode("b", "VP", TreeCoordsOffset.ZERO, setOf("d")),
            UnpositionedBranchingNode("c", "S", TreeCoordsOffset.ZERO, setOf("a", "b")),
        ),
        offset = PlotCoordsOffset.ZERO,
    )

    @Test
    fun positionTerminalNodes() {
        val result = applyNodePositionsToTree(::mockStrWidth, treeWithTerminalNodes)
        assertEquals(CoordsInTree(18.0, -2.0), result.node("a").position)
        assertEquals(CoordsInTree(62.0, -2.0), result.node("b").position)
    }

    @Test
    fun positionTriangleNodes() {
        val result = applyNodePositionsToTree(::mockStrWidth, treeWithTriangleNodes)
        assertEquals(CoordsInTree(18.0, -12.0), result.node("a").position)
        assertEquals(CoordsInTree(79.5, -20.0), result.node("b").position)
    }

    @Test
    fun positionStrandedNodes() {
        val result = applyNodePositionsToTree(::mockStrWidth, treeWithStrandedNodes)
        assertEquals(CoordsInTree(40.0, -42.0), result.node("c").position)
    }

    @Test
    fun positionBranchingNodes() {
        val result = applyNodePositionsToTree(::mockStrWidth, treeWithBranchingNodes)
        assertEquals(CoordsInTree(40.0, -42.0), result.node("c").position)
    }

    @Test
    fun positionBranchingNodesInTreeWithTriangles() {
        val result = applyNodePositionsToTree(::mockStrWidth, treeWithBranchingAndTriangleNodes)
        assertEquals(CoordsInTree(48.75, -60.0), result.node("c").position)
    }

    @Test
    fun positionTriangleVertices() {
        val result = applyNodePositionsToTree(::mockStrWidth, treeWithTriangleNodes)
        val triangleNode = result.nodes["b"] as PositionedTerminalNode
        assertEquals(TreeXRange(40.0, 119.0), triangleNode.triangle)
    }

    @Test
    fun positionVerticallyAlignedNodes() {
        val result = applyNodePositionsToTree(::mockStrWidth, treeWithVerticallyAlignedNode)
        assertEquals(CoordsInTree(18.0, -42.0), result.node("a").position)
    }

    @Test
    fun sortNodesByXCoord() {
        assertContentEquals(
            arrayOf("a", "b"),
            sortNodesByXCoord(::mockStrWidth, treeWithBranchingNodes, setOf("b", "a"))
        )
        assertContentEquals(
            arrayOf("a", "c"),
            sortNodesByXCoord(::mockStrWidth, treeWithBranchingNodes, setOf("a", "c"))
        )
    }

    @Test
    fun measureTreeWidth() {
        val result = applyNodePositionsToTree(::mockStrWidth, treeWithBranchingAndTriangleNodes)
        assertEquals(123.0, result.width)
    }
}
